library/std/src/primitive_docs.rs

```diff
@@ -1527,6 +1527,53 @@ mod prim_ref {}
 /// Note that all of this is not portable to platforms where function pointers and data pointers
 /// have different sizes.
 ///
+/// ### ABI compatibility
+///
+/// Generally, when a function is declared with one signature and called via a function pointer with
+/// a different signature, the two signatures must be *ABI-compatible* or else this call is
+/// Undefined Behavior. ABI compatibility is a lot stricter than merely having the same
+/// representation in memory; for example, even if `i32` and `f32` have the same size and alignment,
+/// they might be passed in different registers and hence not be ABI-compatible.
+///
+/// For two signatures to be considered *ABI-compatible*, they must use a compatible call ABI (as
+/// declared via `extern "ABI"`), must take the same number of arguments, and the individual
+/// argument types and the return types must be ABI-compatible.
+///
+/// The call ABIs are guaranteed to be compatible if they are the same, or if the caller ABI is
+/// `$X-unwind` and the callee ABI is `$X`.
+///
+/// The following types are guaranteed to be ABI-compatible:
+///
+/// - Every type is ABI-compatible with itself.
+/// - `*const T`, `*mut T`, `&T`, `&mut T`, `Box<T>`, `NonNull<T>` are all ABI-compatible with each
+///   other for all `T`.
+/// - Any two `fn()` types with the same `extern` ABI string are ABI-compatible with each other.
+/// - Any two 1-ZST types (types with size 0 and alignment 1) are ABI-compatible.
+/// - A `repr(transparent)` type `T` is ABI-compatible with its unique non-1-ZST field (if there is
+///   such a field).
+/// - `i32` is ABI-compatible with `NonZeroI32`, and similar for all other integer types with their
+///   matching `NonZero*` type.
+/// - If `T` is guaranteed to be subject to the [null pointer
+///   optimization](option/index.html#representation), then `T` and `Option<T>` are ABI-compatible.
+/// - If `T1` and `T2` are ABI-compatible, then two `repr(C)` types that only differ because one
+///   field type was changed from `T1` to `T2` are ABI-compatible.
+/// - ABI-compatibility is symmetric and transitive.
+///
+/// More signatures can be ABI-compatible on specific targets, but that should not be relied upon
+/// since it is not portable and not a stable guarantee.
+///
+/// Noteworthy cases of types *not* being ABI-compatible in general are `bool` vs `u8`, and `i32` vs
+/// `u32`: on some targets, the calling conventions for these types differ in terms of what they
+/// guarantee for the remaining bits in the register that are not used by the value. `i32` vs `f32`
+/// has already been mentioned above.
```

Comment:
its probably best to reiterate that `i32` and `f32` are not ABI-compatible because i had to scroll up to check because i couldn't remember
```suggestion
/// guarantee for the remaining bits in the register that are not used by the value. `i32` and `f32`
/// are not ABI-compatible as mentioned above.
```